### PR TITLE
Add receiver_creator to create receivers at runtime

### PR DESCRIPTION
**Pull request by jrcamp**
This implements a new receiver called receiver_creator. It can start and stop
other types of receivers at runtime. The use for it will be able to discover
services running (such as on Kubernetes or Docker) and be able to start a
receiver to monitor that service. When the service is no longer running its
receiver will also be stopped.

The receiver does not yet react to discovery of services as that is still a
work in progress. This PR is being done as a checkpoint and the
receiver_creator is not yet included in the root go.mod or default components
while it is under heavy development.

Main work to be done:

- Implement service discovery
- Use default factories instead of hard-coded list

### CODE REVIEW

**Comment by owais** [comment]
@jrcamp CircleCI config changed in master and now required the `build-publish` step. Can you please rebase this on latest master and push again? I think that should unblock the PR.

File: receiver/receivercreator/factory.go
Comment (tigrannajaryan) [inline receiver/receivercreator/factory.go]: `receiver.Factory` is renamed to `component.ReceiverFactoryOld`

File: receiver/receivercreator/receiver.go
Comment (tigrannajaryan) [inline receiver/receivercreator/receiver.go]: I submitted https://github.com/open-telemetry/opentelemetry-collector/pull/748
Once it is merged it will be possible to fetch factories.

File: receiver/receivercreator/receiver.go
Comment (tigrannajaryan) [inline receiver/receivercreator/receiver.go]: We typically use `oterr.CombineErrors` when there are multiple errors to return.

File: receiver/receivercreator/factory.go
Comment (jrcamp) [reply receiver/receivercreator/factory.go]: Will update to use new interfaces.

Comment (jrcamp) [reply receiver/receivercreator/factory.go]: Oh looks like I can't update yet because of data package still being internal.

**Comment by jrcamp** [comment]
@tigrannajaryan Updated to address review feedback, will update again when the host factories land.

File: receiver/receivercreator/factory.go
Comment (tigrannajaryan) [reply receiver/receivercreator/factory.go]: Yes, please use the Old interface for now. Once Bogdan is done with new structures we will make them public.

File: go.sum
Comment (tigrannajaryan) [inline go.sum:203]: Please run `make gotidy` to make sure go.sum is tidied up.

File: receiver/receivercreator/config.go
Comment (tigrannajaryan) [inline receiver/receivercreator/config.go:32]: It would be nice to describe what it the key of the map. Even better would be to define this `map[string]interface{}` as a type alias since it is also used elsewhere (e.g. as function parameters).

**Review by tigrannajaryan — COMMENTED** [review]
Generally LGTM, only a couple minor comments left.

Host.GetFactory is already merged to Collector core. You will need update the dependency here so that you can use it.

File: receiver/receivercreator/config.go
Comment (jrcamp) [reply receiver/receivercreator/config.go:32]: I've updated the comment, see if it helps. It's hard to describe the keys since the whole map is arbitrarily defined by the user but I gave it a shot.

File: receiver/receivercreator/config.go
Comment (tigrannajaryan) [inline receiver/receivercreator/config.go:38]: Nit: please use consistent line spacing. Either no empty lines between fields or use empty lines for all.

**Review by tigrannajaryan — APPROVED** [review]
Nice job! Resolve the conflicts and good to merge.

**Comment by jrcamp** [comment]
@tigrannajaryan updated

File: receiver/receivercreator/receiver.go
Comment (bogdandrutu) [inline receiver/receivercreator/receiver.go:66]: Use this:
https://godoc.org/github.com/open-telemetry/opentelemetry-collector/config#NewViper

See also some bugs in Sub https://github.com/open-telemetry/opentelemetry-collector/pull/707

File: receiver/receivercreator/receiver.go
Comment (bogdandrutu) [inline receiver/receivercreator/receiver.go:62]: The names of the arguments are confusing, looking at the calls to this method the last argument has nothing related to "env"

File: receiver/receivercreator/factory.go
Comment (jrcamp) [reply receiver/receivercreator/factory.go:93]: I can't until new data types are moved from internal.

File: receiver/receivercreator/receiver.go
Comment (jrcamp) [reply receiver/receivercreator/receiver.go:154]: Yeah it can pass nextConsumer directly.

File: receiver/receivercreator/receiver.go
Comment (jrcamp) [reply receiver/receivercreator/receiver.go:66]: @bogdandrutu should `viperSub` be exported or just copy the function into here?

File: receiver/receivercreator/receiver.go
Comment (jrcamp) [reply receiver/receivercreator/receiver.go:62]: Changed to discoveredConfig

File: receiver/receivercreator/receiver.go
Comment (bogdandrutu) [reply receiver/receivercreator/receiver.go:66]: You can do a PR to export that temporary.

Comment (jrcamp) [reply receiver/receivercreator/receiver.go]: I updated to just not use sub as it didn't provide much value.

File: receiver/receivercreator/receiver.go
Comment (bogdandrutu) [inline receiver/receivercreator/receiver.go:109]: I am confused: where is this `userConfigMap{}` coming from? Can we just remove from the moment?

File: receiver/receivercreator/factory.go
Comment (bogdandrutu) [inline receiver/receivercreator/factory.go:73]: Can we rely purely on the default "Unmarshaler". Maybe we can define the public "Config" as a raw object (without any custom logic like `receiverType` / `fullName` and have an internalConfig that is created from the raw Config.

Comment (bogdandrutu) [reply receiver/receivercreator/factory.go:73]: Can be done in a separate PR, maybe just file an issue to track if you want.

File: receiver/receivercreator/receiver.go
Comment (jrcamp) [reply receiver/receivercreator/receiver.go:109]: It's how the endpoint will be passed when an instance is discovered. It will be used in the next PR that I have in progress.

File: receiver/receivercreator/factory.go
Comment (jrcamp) [reply receiver/receivercreator/factory.go:73]: Maybe I don't understand the point of CustomUnmarshaler then. We need to be able to unmarshal something like:

```yaml
receiver_creator:
   examplereceiver/1:
       rule: rule text
       config:
            receiverCustomConfig: foo
   ...
```

Is there a better way to do it? The entries under receiver_creator are arbitrary so it has to be a map but I don't think mapstructure can squash a map into the Config struct.

Comment (tigrannajaryan) [reply receiver/receivercreator/factory.go:73]: I think @jrcamp is right. I don't know if there is any other way except to use the CustomUnmarshaler.

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@tigrannajaryan I may have made a mistake, don't know if you were happy with this PR or you had more comments, sorry. Please comment here if there are more comments and we can address them in a separate PR.

**Review by tigrannajaryan — COMMENTED** [review]
@bogdandrutu LGTM :-)

**Pull request merged**
